File: apk_monytix/app/src/main/java/com/example/apk/settings/SettingsViewModel.kt
package com.example.apk.settings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.apk.auth.AuthService
import com.example.apk.config.Config
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException

class SettingsViewModel(
    private val authService: AuthService
) : ViewModel() {
    private val _userId = MutableStateFlow<String?>(null)
    val userId: StateFlow<String?> = _userId.asStateFlow()

    private val _isDeleting = MutableStateFlow(false)
    val isDeleting: StateFlow<Boolean> = _isDeleting.asStateFlow()

    private val _deleteError = MutableStateFlow<String?>(null)
    val deleteError: StateFlow<String?> = _deleteError.asStateFlow()

    private val client = OkHttpClient()

    fun loadUserInfo() {
        viewModelScope.launch {
            try {
                val user = authService.getCurrentUser()
                _userId.value = user?.id
            } catch (e: Exception) {
                android.util.Log.e("SettingsViewModel", "Error loading user info", e)
            }
        }
    }

    fun deleteAllData() {
        viewModelScope.launch {
            _isDeleting.value = true
            _deleteError.value = null

            try {
                val session = authService.getCurrentSession()
                    ?: throw Exception("Not authenticated")

                val request = Request.Builder()
                    .url("${Config.apiBaseUrl}spendsense/data")
                    .delete()
                    .addHeader("Content-Type", "application/json")
                    .addHeader("Authorization", "Bearer ${session.accessToken}")
                    .build()

                val response = client.newCall(request).execute()

                if (!response.isSuccessful) {
                    val errorBody = response.body?.string() ?: "Unknown error"
                    throw IOException("Error ${response.code}: $errorBody")
                }

                // Success
                _deleteError.value = null
                android.util.Log.d("SettingsViewModel", "Successfully deleted all data")
            } catch (e: Exception) {
                android.util.Log.e("SettingsViewModel", "Error deleting data", e)
                _deleteError.value = e.message ?: "Failed to delete data"
            } finally {
                _isDeleting.value = false
            }
        }
    }

    fun clearDeleteError() {
        _deleteError.value = null
    }
}

File: ios_monytix/ios_monytix/settings/SettingsViewModel.swift
//
//  SettingsViewModel.swift
//  ios_monytix
//
//  Created by santosh on 10/01/26.
//

import Foundation
import Combine
import Supabase

@MainActor
class SettingsViewModel: ObservableObject {
    @Published var userId: String?
    @Published var isDeleting = false
    @Published var deleteError: String?
    
    private let authService: AuthService
    private let baseURL: String
    
    init(authService: AuthService) {
        self.authService = authService
        self.baseURL = Config.apiBaseUrl
    }
    
    func loadUserInfo() async {
        if let user = await authService.getCurrentUser() {
            userId = user.id.uuidString
        }
    }
    
    func deleteAllData() async {
        isDeleting = true
        deleteError = nil
        
        do {
            guard let url = URL(string: "\(baseURL)/v1/spendsense/data") else {
                throw SettingsError.invalidURL
            }
            
            var request = URLRequest(url: url)
            request.httpMethod = "DELETE"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            
            // Add auth token
            guard let session = try? await authService.supabase.auth.session else {
                throw SettingsError.notAuthenticated
            }
            request.setValue("Bearer \(session.accessToken)", forHTTPHeaderField: "Authorization")
            
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                throw SettingsError.invalidResponse
            }
            
            guard (200...299).contains(httpResponse.statusCode) else {
                let errorMessage = String(data: data, encoding: .utf8) ?? "Unknown error"
                throw SettingsError.httpError(statusCode: httpResponse.statusCode, message: errorMessage)
            }
            
            // Success
            deleteError = nil
        } catch {
            deleteError = error.localizedDescription
        }
        
        isDeleting = false
    }
}

enum SettingsError: LocalizedError {
    case invalidURL
    case notAuthenticated
    case invalidResponse
    case httpError(statusCode: Int, message: String)
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .notAuthenticated:
            return "Not authenticated"
        case .invalidResponse:
            return "Invalid response"
        case .httpError(let statusCode, let message):
            return "Error \(statusCode): \(message)"
        }
    }
}

